### PR TITLE
Improve QE detection

### PR DESCRIPTION
**Pull request by prckent**
## Proposed changes

Improve user experience by finding and configuring QE automatically if pw.x and pw2qmcpack.x found on PATH. Improve messaging, do not abort if QE not found with QE_BIN specified. Support pw.x and pw2qmcpack.x in different directories (edge case).

Will update docs via update to #2973 when merged.

## What type(s) of changes does this code introduce?

- New feature
- Build related changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

nitrogen, nightly test configurations with cmake 3.19.5

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- NA. Code added or changed in the PR has been clang-formatted
- NA. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: CMake/FindQE.cmake [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# Require both to be present to consider QE_FOUND
+# Take QE_BIN as hint for location
+
+IF(QE_BIN)
+  FIND_PATH(QE_PW2Q_DIR pw2qmcpack.x ${QE_BIN})
+  FIND_PATH(QE_PW_DIR pw.x ${QE_BIN})
+ELSE(QE_BIN)
+  FIND_PATH(QE_PW2Q_DIR pw2qmcpack.x)
+  FIND_PATH(QE_PW_DIR pw.x)
+ENDIF(QE_BIN)
```
Comment (ye-luo) [inline CMake/FindQE.cmake:11]: No need of if(QE_BIN)
The following line allows QE_BIN being empty and if not it is preferred over system PATH.
```
FIND_PATH(QE_PW_DIR pw.x HINTS ${QE_BIN})
FIND_PATH(QE_PW2Q_DIR pw2qmcpack.x HINTS ${QE_BIN})
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -51,12 +51,14 @@ ELSE()
 ENDIF()
   
 ######################################################################
-# Verify QE executables present if QE_BIN specified
-######################################################################
-IF ( DEFINED QE_BIN )
-  INCLUDE(CMake/run_qe.cmake)
-  VERIFY_QE_PRESENT( ${QE_BIN} )
-ENDIF()
+# Check if QE executables available at QE_BIN or on the PATH
```
[the remark this answers is not in the document]
Comment (prckent) [reply CMakeLists.txt:54]: Will do

**Comment by prckent** [comment]
Hmm. We'll be able to delete
```
CHECK_INCLUDE_FILE(dlfcn.h HAVE_DLFCN_H)
CHECK_INCLUDE_FILE(inttypes.h HAVE_INTTYPES_H)
CHECK_INCLUDE_FILE(memory.h HAVE_MEMORY_H)
CHECK_INCLUDE_FILE(sys/stat.h HAVE_SYS_STAT_H)
CHECK_INCLUDE_FILE(sys/time.h HAVE_SYS_TIME_H)
CHECK_INCLUDE_FILE(sys/types.h HAVE_SYS_TYPES_H)
CHECK_INCLUDE_FILE(unistd.h HAVE_UNISTD_H)
```
once blitz is gone. This will speed the build for everyone :-) ( @correaa )

**Review by ye-luo — APPROVED** [review]
I think overall we should keep moving build options towards the beginning of the CMakeLists.txt.
python3 is only needed before subdirectory(src) for unit tests and pyscf/qe should before subdirectory(tests). Keep such changes in the future.

**Comment by prckent** [comment]
A user running cmake is more likely to read the messages if they are printed earlier. It also wouldn't hurt to print a summary table at the end. Hence agreed on the "moving build options up". 

Regarding the python situation, we are repeating a number of tests for module availability and have an unclear situation about which modules are required. I think it would be an improvement to test once for required and optional modules, print a clear summary early in the configuration, and then exclusively use ABC_FOUND flags in the remainder of our cmake. We also need a requirements.txt and a requirements_lite.txt to help automate installs+codify requirements.

**Pull request merged**